### PR TITLE
tools/changelog-helper: Ignore non-MERGED PRs & ignore already listed PRs faster

### PR DESCRIPTION
**Pull request by hoffie**
**Short description of changes**

- tools/changelog-helper: Ignore non-MERGED PRs

- tools/changelog-helper: Ignore already listed PRs faster

CHANGELOG: Internal: Improved Changelog release tooling.

**Context: Fixes an issue?**

https://github.com/jamulussoftware/jamulus/pull/1865#issuecomment-1063468921

**Does this change need documentation? What needs to be documented and how?**

**Status of this Pull Request**

Ready.

**What is missing until this pull request can be merged?**

Ready.

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

**Review by ann0see — APPROVED** [review]

File: tools/changelog-helper.sh [hunk 1]
```diff
@@ -126,10 +126,15 @@ check_or_add_pr() {
         return
     fi
     checked_ids[$id]=1
-    local json=$(gh pr view "${id/#/}" --json title,author)
+    if grep -qF "#$id" <<<"$changelog"; then
```
Comment (pljones) [inline tools/changelog-helper.sh:129]: That will find `#300` if the list contains `#3000` through to `#3009`.  A more precise match would be good.

Comment (hoffie) [reply tools/changelog-helper.sh:129]: Good point. Will probably add `\b` or something around that later.

**Review by pljones — CHANGES_REQUESTED** [review]
Line 129 needs a better matcher.

File: tools/changelog-helper.sh (hunk 1 above)
Comment (hoffie) [reply tools/changelog-helper.sh]: Updated (no \b though).

File: tools/changelog-helper.sh [hunk 2]
```diff
@@ -126,7 +126,7 @@ check_or_add_pr() {
         return
     fi
     checked_ids[$id]=1
-    if grep -qF "#$id" <<<"$changelog"; then
+    if grep -qP "#$id([^0-9]|$)" <<<"$changelog"; then
```
Comment (pljones) [inline tools/changelog-helper.sh:129]: Doesn't `"#$id\>"` work?  I don't think I'd want it matching `#12AB`, either.

Comment (hoffie) [reply tools/changelog-helper.sh]: > Doesn't `"#$id\>"` work?

I didn't know about that -- thanks! It works in `-E` mode (may `\b` might have worked there, too). Updated.

File: tools/changelog-helper.sh [hunk 3]
```diff
@@ -126,10 +126,15 @@ check_or_add_pr() {
         return
     fi
     checked_ids[$id]=1
-    local json=$(gh pr view "${id/#/}" --json title,author)
+    if grep -qE "#$id\>" <<<"$changelog"; then
```
Comment (pljones) [inline tools/changelog-helper.sh:129]: Can you add a comment on what this does and why, please.

Comment (pljones) [reply tools/changelog-helper.sh:129]: And, perhaps - like the "not merged" below - echo a message.

File: tools/changelog-helper.sh (hunk 2 above)
Comment (softins) [reply tools/changelog-helper.sh]: I think `\b` will only work in `-P` (PCRE) mode, if the installed version of `grep` supports it. I prefer `\<` and `\>` if I'm not actually writing perl or PHP.

File: tools/changelog-helper.sh (hunk 3 above)
Comment (hoffie) [reply tools/changelog-helper.sh:129]: > Can you add a comment on what this does and why, please.

Done.

> And, perhaps - like the "not merged" below - echo a message.

Hm, I'm not convinced. It was generally nice to see that repeated invocations of this script added less and less stuff (e.g. when run again later in the release process). If we generate a line of output of each PR even if it was already there, the output becomes harder to read, IMO. If you feel strongly, I can add it though.

Comment (pljones) [reply tools/changelog-helper.sh:129]: Yeah, that's why I was hesitant.  Probably best without.

**Review by pljones — APPROVED** [review]

**Pull request merged**